### PR TITLE
SNOW-1546396: Fix update_resolvable_node to reset attributes for SelectStatement & SetStatement

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1546396

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/compiler/utils.py [hunk 1]
```diff
@@ -193,6 +193,9 @@ def update_resolvable_node(
                           |
                         JOIN
     resolve_node(SelectSnowflakePlan, query_generator) will resolve both SelectSnowflakePlan and SnowflakePlan nodes.
+
+    This operation also resets the previously calculated cumulative_node_complexity for the node so a
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/compiler/utils.py:197]: If we add "reset" operation in this function, this function name is not informational enough - so how about call it `update_resolvable_node_and_reset_complexity`?

**Review by sfc-gh-jdu — APPROVED** [review]

File: src/snowflake/snowpark/_internal/compiler/utils.py (hunk 1 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/utils.py:197]: i think we also reset other recalculable fields like _sql_query, _snowflake_plan, so this name may not be accurate also. maybe we can update the comment to mention that the update includes two parts:
1) Re-Resolve the proper fields based on the child
2) Clear up re-calculatable fields like _sql_query, _snowflake_plan, _cumulative_node_complexity ect

File: src/snowflake/snowpark/_internal/compiler/utils.py [hunk 2]
```diff
@@ -207,20 +210,48 @@ def update_resolvable_node(
             update_resolvable_node(node.source_plan, query_generator)
         resolve_and_update_snowflake_plan(node, query_generator)
 
-    elif isinstance(node, (SelectStatement, SetStatement)):
+    elif isinstance(node, SelectStatement):
+        # clean up the cached sql query and snowflake plan to allow
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/compiler/utils.py:214]: do we need to reset that forSnowflakePlan node also?

Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/utils.py:214]: i think we can simply do it before if isinstance(node, SnowflakePlan) for all type of nodes, right?

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/compiler/utils.py:214]: Snowflake plan actually copies it over from the new resolved snowflake plan which already sets it to None.

**Review by sfc-gh-yzou — APPROVED** [review]

**Pull request merged**